realtime_ai_character/llm/openai_llm.py

```diff
@@ -33,16 +37,38 @@ def __init__(self, model):
                 streaming=True
             )
         self.db = get_chroma()
+        self.search_agent = None
+        if os.getenv('SERPER_API_KEY'):
+            llm = OpenAI(temperature=0)
+            tools = load_tools(["google-serper"], llm=llm)
+            self.search_agent = initialize_agent(
+                tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION
+            )
 
     async def achat(self,
                     history: List[BaseMessage],
                     user_input: str,
                     user_input_template: str,
                     callback: AsyncCallbackTextHandler,
                     audioCallback: AsyncCallbackAudioHandler,
-                    character: Character) -> str:
+                    character: Character,
+                    useSearch: bool=False) -> str:
         # 1. Generate context
         context = self._generate_context(user_input, character)
+        # Get search result if enabled
+        if useSearch:
+            if self.search_agent is None:
+                logger.warning('Search is not enabled, please set SERPER_API_KEY to enable it.')
+            else:
+                search_result: str = self.search_agent.run(character.name + ' ' + user_input)
```

Comment:
qq, is it possible to set limit on how much token it will use?

Reply:
I am little concerned that the search result may take too many tokens, which man confuse LLM

Reply:
I do not think there is such way to limit the token out of box. But according to my experience, since the LLM will summarize the search result first, the result is always pretty short. For example, for the question above in the screenshot, the agent will just reply `Tesla's current stock price is 260.02.` instead of returning all the search result.
<img width="726" alt="image" src="https://github.com/Shaunwei/RealChar/assets/5102398/ac89efe9-9303-4634-a29b-5aeef7b79e93">


Maybe we can add one more step in the chain to limit the words. Maybe we can wait until getting some feedbacks.